pkg/controller/metrics/metrics_with_dynamic_labels.go

```diff
@@ -0,0 +1,177 @@
+		d.fixedLabels = labels
+	}
+}
+
+func WithOptionalLabels(labels map[string]string) func(*dynamicLabels) {
+	return func(d *dynamicLabels) {
+		d.optionalLabels = labels
+	}
+}
+
+// getLabelList combines fixed and optional labels and returns the list without duplicates.
+func (d *dynamicLabels) getLabelList() []string {
+	var finalList []string
+
+	// Ensure there isn't an overlap between fixed and optional labels
+	keys := make(map[string]bool)
+	for _, entry := range append(d.fixedLabels, getKeys(d.optionalLabels)...) {
+		if _, value := keys[entry]; !value {
+			keys[entry] = true
+			finalList = append(finalList, entry)
+		}
+	}
+	return finalList
+}
+
+// buildLabels fetches the optional label values and merges the maps of fixedLabels and optionalLabels without
+// duplicates. The resultant map can be used for observing the metric.
+func (d *dynamicLabels) buildLabels(fixedLabels map[string]string, obj metav1.Object) prometheus.Labels {
+	optionalLabels := make(map[string]string, len(d.optionalLabels))
+	for _, v := range getKeys(d.optionalLabels) {
+		optionalLabels[v] = GetLabelValue(obj, d.optionalLabels[v])
+	}
+	labels := make(map[string]string, len(d.getLabelList()))
+	for label, value := range fixedLabels {
+		labels[label] = value
+	}
+	for label, value := range optionalLabels {
+		labels[label] = value
+	}
+	return labels
+}
+
+// defaultMissingLabels ensures a value is present for all corresponding labels of a metric. If missing, it sets it to a
+// default value
+func defaultMissingLabels(labelKeys []string, valueMap map[string]string) map[string]string {
+	if len(valueMap) < len(labelKeys) {
+		for _, v := range labelKeys {
+			if _, ok := valueMap[v]; !ok {
+				valueMap[v] = constants.MetricLabelDefaultValue
+			}
+		}
+	}
+	return valueMap
+}
+
+type CounterVecWithDynamicLabels struct {
+	*prometheus.CounterOpts
+	metric *prometheus.CounterVec
+	*dynamicLabels
+}
+
+func NewCounterVecWithDynamicLabels(counterOpts *prometheus.CounterOpts,
+	labelOptions ...func(labels *dynamicLabels)) *CounterVecWithDynamicLabels {
+	counterVecMetric := &CounterVecWithDynamicLabels{
+		CounterOpts:   counterOpts,
+		dynamicLabels: &dynamicLabels{},
+	}
+	for _, o := range labelOptions {
+		o(counterVecMetric.dynamicLabels)
+	}
+	counterVecMetric.metric = prometheus.NewCounterVec(*counterVecMetric.CounterOpts, counterVecMetric.getLabelList())
+	return counterVecMetric
+}
+
+// RegisterMetricWithDynamicLabels registers the CounterVec metric.
+func (c CounterVecWithDynamicLabels) RegisterMetricWithDynamicLabels() {
+	metrics.Registry.MustRegister(c.metric)
+}
+
+// ObserveMetricWithDynamicLabels logs the counter metric. It simply increments the value, so last parameter isn't used.
+func (c CounterVecWithDynamicLabels) ObserveMetricWithDynamicLabels(obj metav1.Object, log log.FieldLogger,
+	fixedLabels map[string]string, _ float64) {
+	labels := defaultMissingLabels(c.getLabelList(), c.buildLabels(fixedLabels, obj))
+	c.metric.With(labels).Inc()
```

Comment:
If you take suggestions above, this can just be
```suggestion
	c.metric.With(c.buildLabels(fixedLabels, obj)).Inc()
```
Similar below.